text/0000-node-api-v2.md

```diff
@@ -0,0 +1,994 @@
+    + [Legacy support](#legacy-support)
+    + [API only](#api-only)
+- [Drawbacks](#drawbacks)
+- [Prior art](#prior-art)
+- [Future possibilities](#future-possibilities)
+- [References](#references)
+
+# Summary
+
+[summary]: #summary
+
+Create a v2 JSON-RPC API for the Node API.
+
+# Motivation
+
+[motivation]: #motivation
+
+The current Node API (referred here as v1 API) is a REST API. This API while simple is:
+
+- Documented manually (current documentation is available [here](https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md).
+- Contains call with heterogenous args such as `?byid=xxx` and `commitment/xxx` which can be confusing and lack some uniformity.
+- Uses REST which is bound to HTTP while v2 wallet API uses JSON-RPC.
+
+The goal of this RFC is to provide a new API with:
+- Cleaner methods and errors.
+- Generated documentation directly on docs.rs.
+- Automatic testing with doc tests.
+- Stronger basis for future improvements.
+
+# Community-level explanation
+
+[community-level-explanation]: #community-level-explanation
+
+This new API will be particularly useful for wallet developer and should ultimately simplify their work on Grin. Moreover, this RFC does not introduce any breaking changes as the v1 REST API will still be around until completely deprecated.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Current Endpoints with the v1 API
+
+While the endpoints are documented in details [here](https://github.com/mimblewimble/grin/blob/master/doc/api/node_api.md), here is an overview of the current REST API Endpoints of the V1 API.
+
+```JSON
+[
+  "get blocks",
+  "get headers",
+  "get chain",
+  "post chain/compact",
+  "get chain/validate",
+  "get chain/kernels/xxx?min_height=yyy&max_height=zzz",
+  "get chain/outputs/byids?id=xxx,yyy,zzz",
+  "get chain/outputs/byheight?start_height=101&end_height=200",
+  "get status",
+  "get txhashset/roots",
+  "get txhashset/lastoutputs?n=10",
+  "get txhashset/lastrangeproofs",
+  "get txhashset/lastkernels",
+  "get txhashset/outputs?start_index=1&max=100",
+  "get txhashset/merkleproof?n=1",
+  "get pool",
+  "post pool/push_tx",
+  "post peers/a.b.c.d:p/ban",
+  "post peers/a.b.c.d:p/unban",
+  "get peers/all",
+  "get peers/connected",
+  "get peers/a.b.c.d",
+  "get version"
+]
+```
+
+These endpoints can be grouped into 5 categories:
+
+- miscellaneous endpoints (which contain `status` and `version` endpoints)
+- `chain` endpoints (which also contain `blocks`, `chain` and headers` endpoints)
+- `peer` endpoints
+- `pool` endpoints
+- `txhashset` endpoints
+
+## Proposed Endpoints
+
+This RFC proposes to keep the logic of the following endpoints and refactor them:
+
+```JSON
+[
+  "get blocks",
+  "get headers",
+  "get chain",
+  "post chain/compact",
+  "get chain/validate",
+  "get chain/kernels/xxx?min_height=yyy&max_height=zzz",
+  "get chain/outputs/byids?id=xxx,yyy,zzz",
```

Comment:
Instead of the extra `byids` and `byheight` subresources, you could leverage URL parameters directly on `/chain/outputs`, eg. `/chain/outputs?id=xxx`, `/chain/outputs?min_height=yyy`. Obviously, you would need additional validation to keep the id and height parameters mutually exclusive.

Reply:
See above comment.